core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -186,13 +186,14 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     try {
       val statusList = Option(fs.listStatus(new Path(logDir))).map(_.toSeq)
         .getOrElse(Seq[FileStatus]())
-      var newLastModifiedTime = lastModifiedTime
       val logInfos: Seq[FileStatus] = statusList
         .filter { entry =>
           try {
             getModificationTime(entry).map { time =>
-              newLastModifiedTime = math.max(newLastModifiedTime, time)
-              time >= lastModifiedTime
+              val fileName = entry.getPath.getName
+              val oldLastModifiedTime = lastModifiedTimes.getOrElse(fileName, -1L)
+              lastModifiedTimes += (fileName -> time)
+              time > oldLastModifiedTime
```

Comment:
We can not use `>` to check whether the file is modified. We must use `>=`. Since there might be modification to the file immediately after calling `getModificationTime`, the result of which is the modified time stay the same. We need use `>=` to check the time along with checking the file size changing. For some discussions, please refer to [SPARK-7189](https://issues.apache.org/jira/browse/SPARK-7189)